### ISSUE TITLE
SkinnedMeshRenderer

### ISSUE THREAD

**Issue by ivanpopelyshev** [pixijs/pixi-spine#66]
for objects with many MeshAttachments its crtitical to upload geometry to GPU only for the first time. Every frame upload only bones, mat3 per one bone. 

```javascript
drawArrays(gl.TRIANGLE_STRIP, attachment.startPosition, attachment.endPosition-attachment.startPosition);
```

**Comment by ivanpopelyshev** [comment]
It seems that we need to upload all ffd animations as buffers too. 

Im still not sure if its better to upload bones as texture or as a buffer

**Comment by ivanpopelyshev** [comment]
Work in progress, https://github.com/pixijs/pixi-spine/commits/gpupatch

**Comment by ivanpopelyshev** [comment]
File link: https://github.com/pixijs/pixi-spine/blob/gpupatch/gpupatch/SkinnedMeshPatch.js

**Comment by ivanpopelyshev** [comment]
1) What is best way of integrating it into pixi-spine so it will be possible to switch between vanilla and gpu-mesh version? Best way of configuring pixi plugin?

2) It messes with getLocalBounds and getBounds, new algorithm is calling children.updateTransform() every 200ms if there are no filters and no interaction. It is tested in production, but may be there is another way? WebGL2 will help us to download current vertices back to js.

**Comment by GoodBoyDigital** [comment]
1: Do we even need a vanilla js version if we have the super fast gpu version? I don't think we do :)

2: We can make it so that the user has to define a hit area on the spine model? If we do hittest a mesh, at the very least we should make them super aware that making a mesh interactive is super inefficient. (maybe a warning?)

**Comment by ivanpopelyshev** [comment]
1. Good point. This patch is stable enough, I tested it on on [monsterisland](https://apps.facebook.com/monsterislandonline/) and so far devs are satisfied.

       Theoretically, In case of big number of very simple spine models SpriteRenderer will be faster, and its possible to attach diffirent filters to attachments.

2. Also good point. "interaction and filters are slow unless you specify boundingBox attachment that contains all your animations" :) 

       Oh, I just have an idea: filters can be applied to specific  boundingBoxes. If only spine had polygons for clipping too... We can use meshes for that :)

**Comment by ivanpopelyshev** [comment]
Btw, not related to that pixi-spine:

RpkMaker MV is using PIXI and we have upgrade-to-v3 project going on : 
https://github.com/holywyvern/mv-pixi-upgrade

I separated special tileRenderer (works similar to PIXI.ParticleContainer) into separate library: https://github.com/ivanpopelyshev/pixi-tilemap . I know that vanilla PIXI doesnt use gl.POINTS for that kind of things :)

**Comment by ivanpopelyshev** [comment]
Also, I will move gameofbombs.com to PIXI as soon as I finish all necessary plugins: tilemaps, character animation engine, filters for tilemaps.

**Comment by GoodBoyDigital** [comment]
Exciting stuff man! I love gameofbombs is so awesome. Best way to spend a lunch hour :D

"Just imagine how many japanese hentai games will be made with pixi inside. That certainly deserve to be on pixijs main site." 

haha! that could get er.. interesting!

The tile map is awesome, buttery smooth! - if you want we could move it to the pixi organisation? 

I also see what you mean regarding the spine render modes. If possible could we have it so that it uses sprites, but then uses the GPU for the meshs? 

Thanks again for you hard work here man. This stuff is all fantastic :+1:

**Comment by ivanpopelyshev** [comment]
Please move pixi-tilemap into pixijs org, I'll make docs and benchmarks for it. I'll notify you when I start another library that is marked with "pixi-" prefix.

I forgot that we need vanilla version for canvas fallback, so it wont be a problem to switch between two modes of rendering: vanilla and turbo. 

Thank you! :+1:

**Comment by GoodBoyDigital** [comment]
Good stuff Man! I know lots of people will find the tilemap really handy! (including me!)

@englercj how do we move a project to pixi organisation? Is that something @ivanpopelyshev has to do?

**Comment by englercj** [comment]
@ivanpopelyshev You should have permissions to move it into the organization now.

**Comment by ivanpopelyshev** [comment]
@englercj done! Thank you

**Comment by barbarosso** [comment]
@ivanpopelyshev If i embed the SkinnedMeshRenderer, the spine is rendered two times, but the top one isn't animating.
[preview](http://imgur.com/WAJ9Vks)

**Comment by ivanpopelyshev** [comment]
@barbarosso ugh, i had this bug before. Gonna fix it again :)

**Comment by beanhedge** [comment]
Hi, I get the following error when using with latest pixi-spine:

`SkinnedMeshPatch.js:612 Uncaught TypeError: Cannot read property 'FfdTimeline' of undefined`

Need latest pixi-spine to support sheering.
Need skinnedMeshPatch because I wrote a custom vertex shader.

Thanks!

**Comment by ivanpopelyshev** [comment]
Hi! Tell me please, do you need to use it on 100 skeletons? If not, there's easy solution for custom vertex shader :)

**Comment by beanhedge** [comment]
No, I need like 5 skeletons.  What is the easy solution for custom shader?

**Comment by ivanpopelyshev** [comment]
in v4 its not easy to use shader for sprites/meshes, but its possible: take "pixi-plugin-example" and make a renderer that takes both sprites and meshes. Then specify for every sprite/mesh "mySprite.pluginName='myAwesomeRenderer'"

**Comment by ivanpopelyshev** [comment]
If you want, I can add special setting in spine object that automatically assigns your renderer name in each created sprite/mesh.

**Comment by beanhedge** [comment]
For now I'm OK with using my custom shader globally so I'll give the plug in a shot.  Thanks for the prompt help!

**Comment by ivanpopelyshev** [comment]
Globally -> you have to change spine runtime that way it sets "pluginName=..." for every created thing, right?

**Comment by beanhedge** [comment]
Yeah, it's definitely a hacky solution but it should unblock me.  Your suggested special setting would be nice.

**Comment by amakaseev** [comment]
@ivanpopelyshev 
I welcome, you will continue support of GPU skinning?

**Comment by ivanpopelyshev** [comment]
@amakaseev its still in my todo list and I'm very busy
